### ISSUE TITLE
Losing my committed offsets

### ISSUE THREAD

**Issue by youngpm** [Parsely/pykafka#209]
Hi all,

I have some strange behavior I can't quite get to the bottom of.  I have a consumer listening on a topic in a consumer group like so (I'm using Kafka 0.8.2):

```python
    client = KafkaClient(hosts=kafka_hosts)
    input_topic = client.topics[topic_in]
    consumer = input_topic.get_simple_consumer(consumer_group=group_id,
                                               auto_offset_reset=OffsetType.EARLIEST,
                                               auto_commit_enable=False)

    for message in consumer:
        # Do some work....
        consumer.commit_offsets()
```

This works great most of the time, e.g. I can restart the consumer and it picks up after the last successfully processed message.  

However, I've found that if I restart my consumer after its been idle for quite some time, it occasionally loses its mind and starts processing all the messages in the topic that have already been committed!  It's as if Kafka has reset the offsets on me.

Could it be something to do with where the offsets are being stored (not sure if they're being stored in ZooKeeper or Kafka)?  I came across this guy,

http://kafka.apache.org/documentation.html#offsetmigration

I'm using the defaults for 0.8.2...

Thanks for any tips on getting to the bottom of this!  It's hard for me to reproduce!

**Comment by emmett9001** [comment]
Hi @youngpm, thanks for checking this in. I've actually seen behavior like this before in my local testing, but I've never gotten to the bottom of it. We've never seen it in production, so it hasn't been a super high priority to fix, but it's good to know I'm not the only one who's seen this. Pykafka only knows how to store offsets in kafka itself, so offsets in zookeeper should not be a factor here.

There are two things I recommend changing about your test script to see if/how they affect the results you're seeing. First, I'd use `auto_commit_enable=True` with `auto_commit_interval_ms=1000`. Committing the offset after each message as you're doing should work, but the tests I've run are all using autocommitting since it's slightly more performant in production. Second, I'd change `auto_offset_reset` to `OffsetType.LATEST`. It's possible that your consumer is resetting itself to the beginning of the topic because of this setting, so it would be good to remove that variable.

**Comment by emmett9001** [comment]
This excerpt from the documentation strikes me as a possible cause:

```
In case the offset manager was just started or if it just became the offset manager for a new set of consumer groups (by becoming a leader for a partition of the offsets topic), it may need to load the offsets topic partition into the cache. In this case, the offset fetch will fail with an OffsetsLoadInProgress exception and the consumer may retry the OffsetFetchRequest after backing off.
```

Though the `SimpleConsumer` does handle this case [here](https://github.com/Parsely/pykafka/blob/master/pykafka/simpleconsumer.py#L417).

**Comment by youngpm** [comment]
Hey @emmett9001,

I avoided using ```auto_commit_enable=True``` because my client launches a pretty long running compute job (image processing), and I didn't want the message marked as consumed if it failed.  Same with ```auto_offset_reset``` being set to ```OffsetType.EARLIEST```; I wanted the client to pick up where it left off if it died on me, I think when set to latest you'd miss any messages that came in while you were down.  Might be wrong though, pretty new to Kafka!  Maybe there's a better way to deal with a failure?

I've upped the verbosity on my logs, so next time it happens, maybe I'll have more info!

**Comment by emmett9001** [comment]
Restarting from `EARLIEST` can actually result in repeated messages, since it sets the consumer offset to the earliest offset available for each partition. This means that if you've consumed some messages on a partition and then the consumer resets its offsets, it will go back to the earliest offset still held on that partition - probably one of the ones you just consumed. Neither `EARLIEST` nor `LATEST` will actually guarantee that the consumer will pick up where it left off after a failure - the best way to ensure this is to commit offsets regularly, as you're doing.

**Comment by youngpm** [comment]
I got some more info with a better logging level on!

I had a consumer listening on a topic, no activity on the topic for days.  A message came in, it processed it, but died trying to commit the offsets:

```python
Traceback (most recent call last):
  File "/home/ubuntu/whackstack/metadata_matsel_service.py", line 190, in <module>
    consumer.commit_offsets()
  File "/home/ubuntu/whackstack/venv/local/lib/python2.7/site-packages/pykafka/simpleconsumer.py", line 340, in commit_offsets
    self._consumer_group, 1, 'pykafka', reqs)
  File "/home/ubuntu/whackstack/venv/local/lib/python2.7/site-packages/pykafka/broker.py", line 309, in commit_consumer_group_offsets
    return self._offsets_channel_req_handler.request(req).get(OffsetCommitResponse)
  File "/home/ubuntu/whackstack/venv/local/lib/python2.7/site-packages/pykafka/handlers.py", line 55, in get
    raise self.error
pykafka.exceptions.SocketDisconnectedError
```

This process is running under supervisord, so it restarted and I see this:

```
INFO:2015-07-27 22:12:17,426:cluster.py:_update_brokers:Discovered 1 brokers
INFO:2015-07-27 22:12:17,428:cluster.py:_update_topics:Discovered 15 topics
INFO:2015-07-27 22:12:17,429:topic.py:update:Adding 1 partitions
INFO:2015-07-27 22:12:17,429:topic.py:update:Adding 1 partitions
INFO:2015-07-27 22:12:17,429:topic.py:update:Adding 1 partitions
INFO:2015-07-27 22:12:17,429:topic.py:update:Adding 1 partitions
INFO:2015-07-27 22:12:17,429:topic.py:update:Adding 1 partitions
INFO:2015-07-27 22:12:17,429:topic.py:update:Adding 1 partitions
INFO:2015-07-27 22:12:17,429:topic.py:update:Adding 1 partitions
INFO:2015-07-27 22:12:17,429:topic.py:update:Adding 1 partitions
INFO:2015-07-27 22:12:17,429:topic.py:update:Adding 1 partitions
INFO:2015-07-27 22:12:17,429:topic.py:update:Adding 1 partitions
INFO:2015-07-27 22:12:17,429:topic.py:update:Adding 1 partitions
INFO:2015-07-27 22:12:17,429:topic.py:update:Adding 1 partitions
INFO:2015-07-27 22:12:17,429:topic.py:update:Adding 1 partitions
INFO:2015-07-27 22:12:17,429:topic.py:update:Adding 1 partitions
INFO:2015-07-27 22:12:17,430:cluster.py:get_offset_manager:Attempting to discover offset manager for consumer group 'matsel'
INFO:2015-07-27 22:12:17,432:cluster.py:get_offset_manager:Found coordinator broker with id 1
INFO:2015-07-27 22:12:17,436:simpleconsumer.py:_setup_fetch_workers:Starting 1 fetcher threads
INFO:2015-07-27 22:12:17,438:simpleconsumer.py:_reset_offsets:Resetting offsets for 1 partitions
```

So the last line there says it reset the offsets!  I would guess that the failure to commit the offsets in the first place is related.  The consumer actually processed everything correctly up until the commit_offsets call.  In fact, it posted a message on another topic, where another service died in the same exact way!

That's about all I know, any ideas what would cause the restart?

**Comment by youngpm** [comment]
Just realized I was on version 1.0.3, looks like _reset_offsets doesn't exist anymore.  I've got the latest from pip and will keep monitoring the logs!

**Comment by youngpm** [comment]
@emmett9001, So moving to the latest version on pip (pykafka 1.1.0), I haven't seen any more resetting of offsets!

However, I am now getting quite a number of these fellows when calling ```commit_offsets()``` when the consumer has been idle for a few hours:

```
Traceback (most recent call last):
  File "/home/ubuntu/whackstack/refine_matsel_service.py", line 230, in <module>
    consumer.commit_offsets()
  File "/home/ubuntu/whackstack/venv/local/lib/python2.7/site-packages/pykafka/simpleconsumer.py", line 349, in commit_offsets
    self._consumer_group, 1, 'pykafka', reqs)
  File "/home/ubuntu/whackstack/venv/local/lib/python2.7/site-packages/pykafka/broker.py", line 308, in commit_consumer_group_offsets
    return self._offsets_channel_req_handler.request(req).get(OffsetCommitResponse)
  File "/home/ubuntu/whackstack/venv/local/lib/python2.7/site-packages/pykafka/handlers.py", line 55, in get
    raise self.error
pykafka.exceptions.SocketDisconnectedError
```

The consumer still gets the message and does its thing, but dies at the very end on the commit.  supervisord restarts and it succeeds on the second go around with the commit, not sure what's up!

**Comment by yungchin** [comment]
I wonder if your losing stored offsets previously could be related to the topic having been idle for days, which in combination with your brokers' retention settings may have caused them to get cleaned up by the time you hit this SocketDisconnectedError (which triggered the restart then, too, right?).

**Comment by kbourgoin** [comment]
Good point, @yungchin.  Unlike previous versions of Kafka where offsets were stored forever in Kafka, offset storage rotates after a defined TTL which is `offsets.topic.retention.minutes` in the broker config. [link](http://kafka.apache.org/documentation.html#brokerconfigs).  The default for this is 24h, so if you weren't saving offsets for a few days and it crashed before you saved new values, it makes sense that the old values would be lost.

**Comment by youngpm** [comment]
@kbourgoin Well the offsets were being committed every time a message came in (maybe a few minute delay from getting the message and committing it), its just that no messages were delivered to the topic for a while.  

Basically it was idle for long periods of time in the ```for message in consumer```loop.  I'm not sure if the offset reset is a red herring now, I was on pykafka 1.0.3 when I encountered the offset resets, but the logic has changed in 1.1.0.  Now I'm getting what's described in #224, but I haven't seen a reset of the offsets since I upgraded versions.
